### PR TITLE
core(graph): cleaning `is_graph_kernel`

### PR DESCRIPTION
**Pull request by romintomasetti**
This PR cleans `is_graph_kernel`.

Changes are twofold:
1. For specialized graph implementations, the aggregate node type is now much simpler, because it does not need to define a policy. This change also makes sense, because an aggregate node does not do any work, so it should not have a policy, nor should it be defined as a "graph kernel".
2. `GraphImpl::add_node` is now better constrained. Note that `add_node` is **not** part of the user-facing API, so we don't need to make a nice error message. Also note that I'm strengthening the constraints on this function because I'll introduce a new type of node in a followup PR, requiring `add_node` to be specialized.

### CODE REVIEW

**Review by masterleinad — DISMISSED** [review]
```
/builds/kokkos/kokkos/core/src/SYCL/Kokkos_SYCL_Graph_Impl.hpp:111:26: error: out-of-line definition of 'add_node' does not match any declaration in 'Kokkos::Impl::GraphImpl<Kokkos::SYCL>'
  111 | GraphImpl<Kokkos::SYCL>::add_node(
      |                          ^~~~~~~~
```

File: core/src/impl/Kokkos_Utilities.hpp [hunk 1]
```diff
@@ -120,6 +120,10 @@ struct is_specialization_of : std::false_type {};
 template <template <class...> class Template, class... Args>
 struct is_specialization_of<Template<Args...>, Template> : std::true_type {};
 
+template <typename T, template <typename...> class U>
+inline constexpr bool is_specialization_of_v =
+    is_specialization_of<T, U>::value;
+
```
Comment (JBludau) [inline core/src/impl/Kokkos_Utilities.hpp:126]: I think this should be a separate pr

File: core/src/impl/Kokkos_Default_Graph_Impl.hpp [hunk 2]
```diff
@@ -80,9 +80,9 @@ struct GraphImpl : private ExecutionSpaceInstanceStorage<ExecutionSpace> {
   // <editor-fold desc="required customizations"> {{{2
 
   template <class NodeImpl>
-  //  requires NodeImplPtr is a shared_ptr to specialization of GraphNodeImpl
-  void add_node(std::shared_ptr<NodeImpl> const& arg_node_ptr) {
-    static_assert(NodeImpl::kernel_type::Policy::is_graph_kernel::value);
+  std::enable_if_t<
+      Kokkos::Impl::is_specialization_of_v<NodeImpl, GraphNodeImpl>>
+  add_node(std::shared_ptr<NodeImpl> const& arg_node_ptr) {
```
Comment (JBludau) [inline core/src/impl/Kokkos_Default_Graph_Impl.hpp:85]: hmm ... can we give an easy to read error message to our users? Do you really need SFINAE here?

Comment (romintomasetti) [reply core/src/impl/Kokkos_Default_Graph_Impl.hpp:85]: SFINAE is not 100% needed now, but a following PR will introduce a new type of node, for which we'll need to specialize `add_node`;

File: core/src/impl/Kokkos_Utilities.hpp (hunk 1 above)
Comment (romintomasetti) [reply core/src/impl/Kokkos_Utilities.hpp:126]: Given that `is_specialization_of` is only used in the graph stuff, I would say that it's OK not to make a separate PR with just this change. Up to you.

Comment (JBludau) [reply core/src/impl/Kokkos_Utilities.hpp:126]: yeah, it is probably fine

File: core/src/impl/Kokkos_Default_Graph_Impl.hpp (hunk 2 above)
Comment (JBludau) [reply core/src/impl/Kokkos_Default_Graph_Impl.hpp:85]: ah ok. well, if users never hit this I guess it is fine.

File: core/src/impl/Kokkos_Utilities.hpp (hunk 1 above)
Comment (romintomasetti) [reply core/src/impl/Kokkos_Utilities.hpp:126]: Thanks, closing then.

File: core/src/impl/Kokkos_Default_Graph_Impl.hpp (hunk 2 above)
Comment (romintomasetti) [reply core/src/impl/Kokkos_Default_Graph_Impl.hpp:85]: @JBludau i updated the PR description. If it's OK for you, please close this thread :wink:

**Comment by romintomasetti** [comment]
@masterleinad @JBludau I don't know if we care about the Windows build failure. It seems unrelated to my changes, isn't ?

File: core/src/Cuda/Kokkos_Cuda_Graph_Impl.hpp [hunk 3]
```diff
@@ -121,10 +121,10 @@ struct GraphImpl<Kokkos::Cuda> {
...
 
   template <class NodeImpl>
-  //  requires NodeImplPtr is a shared_ptr to specialization of GraphNodeImpl
-  //  Also requires that the kernel has the graph node tag in its policy
-  void add_node(std::shared_ptr<NodeImpl> const& arg_node_ptr) {
-    static_assert(NodeImpl::kernel_type::Policy::is_graph_kernel::value);
+  std::enable_if_t<
+      Kokkos::Impl::is_specialization_of_v<NodeImpl, GraphNodeImpl> &&
+      Kokkos::Impl::is_graph_kernel_v<typename NodeImpl::kernel_type>>
+  add_node(std::shared_ptr<NodeImpl> const& arg_node_ptr) {
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply core/src/Cuda/Kokkos_Cuda_Graph_Impl.hpp:127]: Then, it seems a `static_assert` for `is_specialization_of_v<NodeImpl, GraphNodeImpl` instead using SFINAE for it makes sense.

[the remark this answers is not in the document]
Comment (romintomasetti) [reply core/src/Cuda/Kokkos_Cuda_Graph_Impl.hpp]: Indeed, and done :heavy_check_mark:

**Review by masterleinad — APPROVED** [review]
Fine with me.

File: core/src/Cuda/Kokkos_Cuda_Graph_Impl.hpp [hunk 4]
```diff
@@ -121,10 +121,11 @@ struct GraphImpl<Kokkos::Cuda> {
   }
 
   template <class NodeImpl>
-  //  requires NodeImplPtr is a shared_ptr to specialization of GraphNodeImpl
-  //  Also requires that the kernel has the graph node tag in its policy
-  void add_node(std::shared_ptr<NodeImpl> const& arg_node_ptr) {
-    static_assert(NodeImpl::kernel_type::Policy::is_graph_kernel::value);
+  std::enable_if_t<
+      Kokkos::Impl::is_graph_kernel_v<typename NodeImpl::kernel_type>>
```
Comment (dalg24) [inline core/src/Cuda/Kokkos_Cuda_Graph_Impl.hpp:125]: I am confused.  Where is that `Impl::is_graph_kernel_v` helper variable template defined?

**Review by dalg24 — CHANGES_REQUESTED** [review]

File: core/src/Cuda/Kokkos_Cuda_Graph_Impl.hpp (hunk 4 above)
Comment (romintomasetti) [reply core/src/Cuda/Kokkos_Cuda_Graph_Impl.hpp:125]: Not easy to find indeed.

https://github.com/kokkos/kokkos/blob/bc6581798c5efe1abc8b1b937baa7f2185414665/core/src/Kokkos_Concepts.hpp#L168

**Review by dalg24 — APPROVED** [review]
Summarizing the changes here:

* Turn the `GraphImpl::add_node()` **mandate** (i.e. hard requirement on type enforced via a static assertion) that the node pointer argument is tagged as a graph kernel into a **constraint** (i.e. control on overload resolution) in preparation of adding a new overload that takes some other kind of graph node
* Get rid of the dummy `Policy` "tag" on aggregate nodes
* Add `Impl::is_specialization_v` helper variable template
